### PR TITLE
Cb 4198 gcp file system support

### PR DESCRIPTION
**Pull request by yagudin10**

### CODE REVIEW

File: server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/WebFSUtils.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.jkiss.dbeaver.model.navigator.DBNModel;
+import org.jkiss.dbeaver.model.navigator.DBNNode;
+import org.jkiss.dbeaver.model.navigator.fs.DBNPathBase;
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.text.MessageFormat;
+
+public class WebFSUtils {
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/WebFSUtils.java]: use FSUtils

File: server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java [hunk 2]
```diff
@@ -146,13 +150,18 @@ public FSFile writeFileContent(
...
         try {
-            Path filePath = webSession.getFileSystemManager(projectId).getPathFromURI(webSession.getProgressMonitor(), fileUri);
+            DBNPathBase parentNode = WebFSUtils.getNodeByPath(webSession, parentPath);
+            if (!Files.isDirectory(parentNode.getPath())) {
+                throw new DBException(MessageFormat.format("Node ''{0}'' is not a directory", parentPath));
+            }
+            Path filePath = parentNode.getPath().resolve(fileName);
             Files.createFile(filePath);
-            return new FSFile(filePath);
+            parentNode.refreshNode(webSession.getProgressMonitor(), this);
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java]:
```suggestion
            parentNode.addChildResource(filePath);
            return new FSFile(parentNode.getChild(filePath));
```

File: server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java [hunk 3]
```diff
@@ -161,15 +170,37 @@ public FSFile createFile(
...
+    ) throws DBWebException {
+        try {
+            Path from = WebFSUtils.getPathFromNode(webSession, oldNodePath);
+            Path fileName = from.getFileName();
+            DBNPathBase parentNode = WebFSUtils.getNodeByPath(webSession, parentNodePath);
+            Path parentPath = parentNode.getPath();
+            if (!Files.isDirectory(parentPath)) {
+                throw new DBException(MessageFormat.format("Node ''{0}'' is not a directory", parentPath));
+            }
+            Path to = Files.move(from, parentPath.resolve(fileName));
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java]: same here

File: server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java [hunk 4]
```diff
@@ -161,15 +170,37 @@ public FSFile createFile(
...
     ) throws DBWebException {
         try {
-            Path from = webSession.getFileSystemManager(projectId).getPathFromURI(webSession.getProgressMonitor(), fromURI);
-            Path to = webSession.getFileSystemManager(projectId).getPathFromURI(webSession.getProgressMonitor(), toURI);
-            Files.move(from, to);
-            return new FSFile(to);
+            DBNPathBase dbnPath = WebFSUtils.getNodeByPath(webSession, nodePath);
+            Path oldpath = dbnPath.getPath();
+            Path newPath = Files.move(oldpath, oldpath.resolveSibling(newName));
+            dbnPath.refreshNode(webSession.getProgressMonitor(), this);
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java]:
```suggestion
            dbnPath.removeChildResource(oldpath);
            dbnPath.addChildResource(newPath);```

File: server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java [hunk 5]
```diff
@@ -179,13 +210,18 @@ public FSFile moveFile(
...
         try {
-            Path folderPath = webSession.getFileSystemManager(projectId).getPathFromURI(webSession.getProgressMonitor(), folderURI);
+            DBNPathBase parentNode = WebFSUtils.getNodeByPath(webSession, parentPath);
+            if (!Files.isDirectory(parentNode.getPath())) {
+                throw new DBException(MessageFormat.format("Node ''{0}'' is not a directory", parentPath));
+            }
+            Path folderPath = parentNode.getPath().resolve(folderName);
             Files.createDirectory(folderPath);
-            return new FSFile(folderPath);
+            parentNode.refreshNode(webSession.getProgressMonitor(), this);
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.fs/src/io/cloudbeaver/service/fs/impl/WebServiceFS.java]: same here

**Review by alexander-skoblikov — CHANGES_REQUESTED** [review]

**Review by devnaumov — APPROVED** [review]

**Review by alexander-skoblikov — APPROVED** [review]

**Pull request merged**